openresty-patches/patches/1.21.4.1/ngx_lua-0.10.21_02-dyn_upstream_keepalive.patch

```diff
@@ -617,63 +643,78 @@ index f71a3e00..0d403716 100644
+ 
+-    if (rc != 0) {
+-        /*  error occurred when running loaded code */
+-        err_msg = (u_char *) lua_tolstring(L, -1, &len);
+ 
+-        if (err_msg == NULL) {
+-            err_msg = (u_char *) "unknown reason";
+-            len = sizeof("unknown reason") - 1;
+-        }
 +static ngx_int_t
 +ngx_http_lua_balancer_create_keepalive_pool(lua_State *L, ngx_log_t *log,
-+    uint32_t cpool_crc32, ngx_uint_t cpool_size,
++    ngx_str_t *cpool_name, ngx_uint_t cpool_size,
 +    ngx_http_lua_balancer_keepalive_pool_t **cpool)
 +{
 +    size_t                                       size;
 +    ngx_uint_t                                   i;
 +    ngx_http_lua_balancer_keepalive_pool_t      *upool;
 +    ngx_http_lua_balancer_keepalive_item_t      *items;
-+
+ 
+-        ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
+-                      "failed to run balancer_by_lua*: %*s", len, err_msg);
 +    /* get upstream connection pools table */
 +    lua_pushlightuserdata(L, ngx_http_lua_lightudata_mask(
 +                          balancer_keepalive_pools_table_key));
 +    lua_rawget(L, LUA_REGISTRYINDEX); /* pools? */
-+
+ 
+-        lua_settop(L, 0); /*  clear remaining elems on stack */
 +    ngx_http_lua_assert(lua_istable(L, -1));
 +
-+    size = sizeof(ngx_http_lua_balancer_keepalive_pool_t)
-+           + sizeof(ngx_http_lua_balancer_keepalive_item_t) * cpool_size;
-
-+    upool = lua_newuserdata(L, size); /* pools upool */
++    lua_pushlstring(L, (const char *)cpool_name->data, cpool_name->len);
+ 
++    size = sizeof(ngx_http_lua_balancer_keepalive_pool_t) +
++           sizeof(ngx_http_lua_balancer_keepalive_item_t) * cpool_size;
++
++    upool = lua_newuserdata(L, size + cpool_name->len); /* pools upool */
 +    if (upool == NULL) {
          return NGX_ERROR;
      }
-
+ 
 -    lua_settop(L, 0); /*  clear remaining elems on stack */
 -    return rc;
 +    ngx_log_debug2(NGX_LOG_DEBUG_HTTP, log, 0,
-+                   "lua balancer: keepalive create pool, crc32: %ui, "
-+                   "size: %ui", cpool_crc32, cpool_size);
++                   "lua balancer: keepalive create pool, "
++                   "name: %V, size: %ui",
++                   cpool_name, cpool_size);
 +
 +    upool->lua_vm = L;
-+    upool->crc32 = cpool_crc32;
 +    upool->size = cpool_size;
 +    upool->connections = 0;
 +
++    upool->cpool_name.len = cpool_name->len;
++    upool->cpool_name.data = (u_char *)(upool) + size;
++    ngx_memcpy(upool->cpool_name.data, cpool_name->data, cpool_name->len);
++
 +    ngx_queue_init(&upool->cache);
 +    ngx_queue_init(&upool->free);
 +
-+    lua_rawseti(L, -2, cpool_crc32); /* pools */
++    lua_rawset(L, -3); /* pools */
 +    lua_pop(L, 1); /* orig stack */
 +
 +    items = (ngx_http_lua_balancer_keepalive_item_t *) (&upool->free + 1);
 +
-+    ngx_http_lua_assert((void *) items == ngx_align_ptr(items, NGX_ALIGNMENT));
++    /*ngx_http_lua_assert((void *) items == ngx_align_ptr(items, NGX_ALIGNMENT));*/
```

Comment:
`lua_newuserdata` may not allocate aligned memory, so this assert will fail occasionally.